File: haushaltsbuch/src/main/java/org/github/haschi/haushaltsbuch/modell/Haushaltsbuch.java
package org.github.haschi.haushaltsbuch.modell;

import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.commandhandling.model.AggregateIdentifier;
import org.axonframework.commandhandling.model.AggregateLifecycle;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.github.haschi.haushaltsbuch.api.BeginneHaushaltsbuchführung;
import org.github.haschi.haushaltsbuch.api.Buchung;
import org.github.haschi.haushaltsbuch.api.Eröffnungsbilanzkonto;
import org.github.haschi.haushaltsbuch.api.EröffnungsbilanzkontoErstellt;
import org.github.haschi.haushaltsbuch.api.HaushaltsbuchführungBegonnen;
import org.github.haschi.haushaltsbuch.api.Inventar;
import org.github.haschi.haushaltsbuch.api.Währungsbetrag;
import org.github.haschi.haushaltsbuch.infrastruktur.modellierung.de.Aggregatkennung;

public final class Haushaltsbuch
{

    @AggregateIdentifier
    private Aggregatkennung id;

    public Haushaltsbuch()
    {
    }

    @CommandHandler
    public Haushaltsbuch(final BeginneHaushaltsbuchführung anweisung)
    {
        AggregateLifecycle.apply(
                HaushaltsbuchführungBegonnen.builder()
                        .id(anweisung.id())
                        .build());

        final Inventar inventar = anweisung.inventar();

        final Währungsbetrag eigenkapital = Währungsbetrag.of(
                inventar.anlagevermoegen().summe().wert()
                        .add(inventar.umlaufvermoegen().summe().wert())
                        // .subtract(inventar.schulden().summe().wert())
        );

        final Eröffnungsbilanzkonto eröffnungsbilanz = Eröffnungsbilanzkonto.builder()
                .addHaben(Buchung.builder()
                                  .buchungstext("Anlagevermögen (AV)")
                                  .betrag(inventar.anlagevermoegen().summe())
                                  .build())
                .addHaben(Buchung.builder()
                                  .buchungstext("Umlaufvermögen (UV)")
                                  .betrag(inventar.umlaufvermoegen().summe())
                                  .build())
                .addSoll(Buchung.builder()
                                 .buchungstext("Eigenkapital (EK)")
                                 .betrag(eigenkapital)
                                 .build())
//                .addSoll(Buchung.builder()
//                                 .buchungstext("Fremdkapital (FK)")
//                                 .betrag(inventar.schulden().summe())
//                                 .build())
                .build();

        AggregateLifecycle.apply(
                EröffnungsbilanzkontoErstellt.builder()
                        .eröffnungsbilanzkonto(eröffnungsbilanz)
                        .build());
    }

    @EventSourcingHandler
    public void falls(final HaushaltsbuchführungBegonnen ereignis)
    {
        id = ereignis.id();
    }
}

File: haushaltsbuch/src/main/kotlin/org/github/haschi/haushaltsbuch/modell/Haushaltsbuch.kt
package org.github.haschi.haushaltsbuch.modell

import org.axonframework.commandhandling.CommandHandler
import org.axonframework.commandhandling.model.AggregateIdentifier
import org.axonframework.commandhandling.model.AggregateLifecycle
import org.axonframework.eventsourcing.EventSourcingHandler
import org.github.haschi.haushaltsbuch.modell.core.commands.BeginneHaushaltsbuchführung
import org.github.haschi.haushaltsbuch.modell.core.events.EröffnungsbilanzkontoErstellt
import org.github.haschi.haushaltsbuch.modell.core.events.HaushaltsbuchführungBegonnen
import org.github.haschi.haushaltsbuch.modell.core.values.Buchung
import org.github.haschi.haushaltsbuch.modell.core.values.Eröffnungsbilanzkonto
import org.github.haschi.haushaltsbuch.modell.core.values.Währungsbetrag
import org.github.haschi.haushaltsbuch.modell.core.values.Aggregatkennung

class Haushaltsbuch {

    @AggregateIdentifier
    private var id: Aggregatkennung? = null

    constructor()

    @CommandHandler
    constructor(anweisung: BeginneHaushaltsbuchführung) {
        AggregateLifecycle.apply(
                HaushaltsbuchführungBegonnen(anweisung.id))

        val inventar = anweisung.inventar

        val eigenkapital = Währungsbetrag(
                inventar.anlagevermoegen.summe.wert
                        .add(inventar.umlaufvermoegen.summe.wert)
                // .subtract(inventar.schulden().summe().wert())
        )

        val eröffnungsbilanz = Eröffnungsbilanzkonto(
                haben = listOf(
                        Buchung(
                                buchungstext = "Anlagevermögen (AV)",
                                betrag = inventar.anlagevermoegen.summe),
                        Buchung(
                                buchungstext = "Umlaufvermögen (UV)",
                                betrag = inventar.umlaufvermoegen.summe)),
                soll = listOf(
                        Buchung(
                                buchungstext = "Eigenkapital (EK)",
                                betrag = eigenkapital)))

        //                .addSoll(Buchung.builder()
        //                                 .buchungstext("Fremdkapital (FK)")
        //                                 .betrag(inventar.schulden().summe())
        //                                 .build())


        AggregateLifecycle.apply(
                EröffnungsbilanzkontoErstellt(eröffnungsbilanz))
    }

    @EventSourcingHandler
    fun falls(ereignis: HaushaltsbuchführungBegonnen) {
        id = ereignis.id
    }
}
